library/core/src/ptr/non_null.rs

```diff
@@ -1472,10 +1476,17 @@ impl<T> NonNull<[T]> {
     ///
     /// (Note that this example artificially demonstrates a use of this method,
     /// but `let slice = NonNull::from(&x[..]);` would be a better way to write code like this.)
+    // Safety: https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html
     #[stable(feature = "nonnull_slice_from_raw_parts", since = "1.70.0")]
     #[rustc_const_stable(feature = "const_slice_from_raw_parts_mut", since = "1.83.0")]
     #[must_use]
     #[inline]
+    #[requires(data.pointer.is_aligned() 
+        && len as isize * core::mem::size_of::<T>() as isize <= isize::MAX
+        && (len as isize).checked_mul(core::mem::size_of::<T>() as isize).is_some()
+        && (data.pointer as isize).checked_add(len as isize * core::mem::size_of::<T>() as isize).is_some() // adding len must not “wrap around” the address space
+        && unsafe { kani::mem::same_allocation(data.pointer, data.pointer.add(len)) })] 
```

Comment:
Another safety requirement is that `data` must be valid for reads. This can be specified using `can_dereference`.